serde/src/de/buffer.rs

```diff
@@ -0,0 +1,1809 @@
+//! Data structures for buffering self-describing formats.
+//!
+//! ```
+//! # use serde::de::{Deserialize, value, IntoDeserializer, buffer::Buffer};
+//! # let source = value::U64Deserializer::<value::Error>::new(32);
+//! // Deserialize any self-describing format from `source` into `buffer`.
+//! let buffer = Buffer::deserialize(source).unwrap();
+//! // Turn the buffer back into a deserializer.
+//! let deserializer = IntoDeserializer::<value::Error>::into_deserializer(buffer);
+//! # assert_eq!(u32::deserialize(deserializer).unwrap(), 32);
+//! ```
+
+use lib::*;
+
+use __private::size_hint;
+
+use de::{self, Deserialize, Deserializer, EnumAccess, Expected, MapAccess, SeqAccess, Visitor};
+
+/// An efficient buffer for self-describing formats.
+#[derive(Debug, Clone)]
+pub struct Buffer<'de>(pub(crate) BufferInner<'de>);
+
+impl<'de> Buffer<'de> {
+    /// Tries to extract a string from the buffer.
+    ///
+    // TODO: This function is used from within code generated by `serde_derive` and, thus,
+    // needs to be public. If this is undesired, we can could simply have it as a normal
+    // function in `private::de` instead.
+    pub fn as_str(&self) -> Option<&str> {
+        match self.0 {
+            BufferInner::Str(x) => Some(x),
+            BufferInner::String(ref x) => Some(x),
+            BufferInner::Bytes(x) => str::from_utf8(x).ok(),
+            BufferInner::ByteBuf(ref x) => str::from_utf8(x).ok(),
+            _ => None,
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Buffer<'de> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let visitor = BufferVisitor::new();
+        deserializer.__deserialize_buffer(::actually_private::T, visitor)
+    }
+}
+
+impl<'de, E: de::Error> de::IntoDeserializer<'de, E> for Buffer<'de> {
+    type Deserializer = BufferDeserializer<'de, E>;
+
+    fn into_deserializer(self) -> Self::Deserializer {
+        BufferDeserializer {
+            buffer: self,
+            err: PhantomData,
+        }
+    }
+}
+
+impl<'a, 'de, E: de::Error> de::IntoDeserializer<'de, E> for &'a Buffer<'de> {
+    type Deserializer = BufferRefDeserializer<'a, 'de, E>;
+
+    fn into_deserializer(self) -> Self::Deserializer {
+        BufferRefDeserializer::new(self)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum BufferInner<'de> {
+    Bool(bool),
+
+    U8(u8),
+    U16(u16),
+    U32(u32),
+    U64(u64),
```

Comment:
Is it possible to extend the support to `u128` and `i128`. Not all old compiler versions support it, but it is supported on all targets starting from Rust 1.40. Given that this is now a private type, that should not be a problem, as all the depending code can also be cfg-gated.

Reply:
Good point! I modified the PR accordingly.